### PR TITLE
Basic support for `scipy.special` v2

### PR DESCRIPTION
**Pull request by person142**
Continuation of https://github.com/numba/numba-scipy/pull/12. Opened a new PR because the code is reasonably different. On a high level, things still work the same way:

- Generate a big list of `(<function_name>, <numba_signature>) -> <function_pointer>`
  - The script that does that is `special_function_pointers.py`
  - The generated code is in `special.signatures.py`
- At import time, iterate over all those signatures and add overloads.

The big change is that the function pointers are grabbed directly from `__pyx_capi__`. This necessitates some de-mangling of Cython function names and parsing of capsule names, but it is all kept to the codegen phase (i.e. none of it is happening at run time).

### CODE REVIEW

**Comment by person142** [comment]
Ok, the remaining test failures are interesting; most of them are small differences in the return values of the function; one for `eval_jacobi` is a `nan`/`inf` mismatch.

The `nan`/`inf` mismatch is likely a SciPy bug; there's a reasonable chance that some of the others are too. I'll note that on the SciPy side we've been only loosely comparing the results of `special` and `cython_special` when testing:

https://github.com/scipy/scipy/blob/master/scipy/special/tests/test_cython_special.py#L331

which we might have to resort to here.

File: numba_scipy/special/types.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+import numba
+
+bytes_to_int_type = {
+    2: numba.types.int16,
+    4: numba.types.int32,
+    8: numba.types.int64,
+}
+
+numba_long = bytes_to_int_type[ctypes.sizeof(ctypes.c_long)]
```
[the remark this answers is not in the document]
Comment (person142) [reply numba_scipy/special/types.py]: Oof I missed that it exists!

File: numba_scipy/special/signatures.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
+import ctypes
+
+import numba
+from numba.extending import get_cython_function_address
+
+from .types import numba_long
+
+name_to_numba_signatures = {
```
[the remark this answers is not in the document]
Comment (person142) [reply numba_scipy/special/signatures.py]: That’s a pretty pressing concern-I added a function recently here https://github.com/scipy/scipy/pull/10440 and have plans to add a few more.

**Comment by person142** [comment]
Thanks for reviewing! I’ll fix things up over the weekend. Re

> There's a few of warnings in the tests

The story is that originally some Cephes  functions took floats as arguments and silently rounded them. SciPy updated them to take ints, but preserved the old legacy rounding behavior for back-compat and added the warnings we’re seeing in the tests.

tl;dr version is that it’s safe to silence them on the numba-scipy side.

**Comment by stuartarchibald** [comment]
> Thanks for reviewing! I’ll fix things up over the weekend. Re

No problem, thanks for doing the work!!

> 
> > There's a few of warnings in the tests
> 
> The story is that originally some Cephes functions took floats as arguments and silently rounded them. SciPy updated them to take ints, but preserved the old legacy rounding behavior for back-compat and added the warnings we’re seeing in the tests.
> 
> tl;dr version is that it’s safe to silence them on the numba-scipy side.

Thanks for the back story, it seemed like something along those lines must have gone on. Agree that silencing them in the tests would be an appropriate choice here.

File: numba_scipy/special/signatures.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    'spence': [(numba.types.float64,)],
+    'xlog1py': [(numba.types.float64, numba.types.float64,)],
+    'xlogy': [(numba.types.float64, numba.types.float64,)],
+    'yn': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'yv': [(numba.types.float64, numba.types.float64,)],
+    'yve': [(numba.types.float64, numba.types.float64,)]
+}
+
+name_and_types_to_pointer = {
+    ('agm', numba.types.float64, numba.types.float64): ctypes.CFUNCTYPE(ctypes.c_double, ctypes.c_double, ctypes.c_double)(get_cython_function_address('scipy.special.cython_special', 'agm')),
```
[the remark this answers is not in the document]
Comment (person142) [reply numba_scipy/special/signatures.py]: I've been testing it with this script:

```python
import timeit
import ctypes

import scipy.special as sc
import numba
import numba_scipy.special

expit_address = numba.extending.get_cython_function_address(
    'scipy.special.cython_special',
    '__pyx_fuse_0expit',
)

expit_pointer = ctypes.CFUNCTYPE(
    ctypes.c_double,
    ctypes.c_double,
    ctypes.c_int,
)(expit_address)

expit_pointer_no_dispatch = ctypes.CFUNCTYPE(
    ctypes.c_double,
    ctypes.c_double,
)(expit_address)


@numba.njit
def expit_dispatch_is_1():
    for i in range(10**4):
        expit_pointer(2.0, 1)


@numba.njit
def expit_dispatch_is_0():
    for i in range(10**4):
        expit_pointer(2.0, 0)


@numba.njit
def expit_dispatch_not_specified():
    for i in range(10**4):
        expit_pointer_no_dispatch(2.0)


@numba.njit
def expit_numba_scipy():
    for i in range(10**4):
        sc.expit(2.0)


if __name__ == '__main__':
    print(timeit.timeit(expit_dispatch_is_1, number=100))
    print(timeit.timeit(expit_dispatch_is_0, number=100))
    print(timeit.timeit(expit_dispatch_not_specified, number=100))
    print(timeit.timeit(expit_numba_scipy, number=100))
```

which on my machine gives this:

```
0.07619187600000005
0.039562226000000034
0.036896394
0.049485624000000006
```

So it seems like ignoring it is fine? Not sure why the `numba_scipy` version is consistently a little slower though. Maybe immediately after this I should set up some asv benchmarks so that we can try to nail down what's happening more precisely.

**Comment by person142** [comment]
Ok, I think I've addressed the review comments.

File: numba_scipy/special/signatures.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba_scipy/special/signatures.py]: I was more worried about correctness, but this indeed is a good point. I think the above code is including compilation time in the timings, hence the first function being quite expensive as numba compilation machinery is initialised for that one, and then the rest benefit from it. The rest of the functions will also contain some compilation cost which is likely why their timings are a bit bumpy. This:
```python
import timeit
import ctypes

import scipy.special as sc
import numba
import numba_scipy.special

expit_address = numba.extending.get_cython_function_address(
    'scipy.special.cython_special',
    '__pyx_fuse_0expit',
)

expit_pointer = ctypes.CFUNCTYPE(
    ctypes.c_double,
    ctypes.c_double,
    ctypes.c_int,
)(expit_address)

expit_pointer_no_dispatch = ctypes.CFUNCTYPE(
    ctypes.c_double,
    ctypes.c_double,
)(expit_address)


@numba.njit
def expit_dispatch_is_1():
    for i in range(10**4):
        expit_pointer(2.0, 1)


@numba.njit
def expit_dispatch_is_0():
    for i in range(10**4):
        expit_pointer(2.0, 0)


@numba.njit
def expit_dispatch_not_specified():
    for i in range(10**4):
        expit_pointer_no_dispatch(2.0)


@numba.njit
def expit_numba_scipy():
    for i in range(10**4):
        sc.expit(2.0)


if __name__ == '__main__':
    # warm up
    expit_dispatch_is_1()
    expit_dispatch_is_0()
    expit_dispatch_not_specified()
    expit_numba_scipy()

    print(timeit.timeit(expit_dispatch_is_1, number=100))
    print(timeit.timeit(expit_dispatch_is_0, number=100))
    print(timeit.timeit(expit_dispatch_not_specified, number=100))
    print(timeit.timeit(expit_numba_scipy, number=100))
```
produces near consistent times for me.

**Comment by stuartarchibald** [comment]
Thanks for making the fixes, this looks good. Merging!

File: .gitattributes [hunk 4]
```diff
@@ -1 +1,2 @@
 numba_scipy/_version.py export-subst
+numba_scipy/special/signatures.py binary
```
Comment (esc) [inline .gitattributes:2]: @person142 -- do you happen to have any idea why this change was made? We are currently proposing to remove it again in #54 because we don't see a need for it and can't see why it would be necessary.

Comment (person142) [reply .gitattributes:2]: No idea. Probably no Chesterton’s fence here.
